### ISSUE TITLE
Jersey Multipart Form Data File Upload functionality appears broken

### ISSUE THREAD

**Issue by seshuad** [awslabs/aws-serverless-java-container#223]
* Implementations: Jersey 

## Expected behavior
ServletInputStream size should match the size of the file being uploaded us Mulitpart Form upload. 

## Actual behavior
The ServletInputStream size does not match the size of the file being Uploaded. AwsProxyRequest maps the request body to a String which might be the cause. 

@JsonIgnoreProperties(ignoreUnknown = true)
public class AwsProxyRequest {

    private String body;
    private String resource;

## Steps to reproduce
Implement a Jersey Lamda Function with @FormDataParam for e.g. 

   public Response uploadFile(
            @FormDataParam("file") InputStream uploadedInputStream,
            @FormDataParam("file") FormDataContentDisposition fileDetail)

**Comment by sapessi** [comment]
Hey @seshuad, the input model maps to a stream because that is indeed how Lambda receives the data - Lambda only supports JSON events so API Gateway and ALB automatically base64 encode the body before sending it to Lambda. The framework decodes the data before sending it to your implementation. 

In API Gateway, you need to [explicitly enable binary support](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-payload-encodings.html) for the content types you want. I'll try to replicate this issue over the next couple of days.

**Comment by sapessi** [comment]
@seshuad can you tell me what sizes you see?

```java
System.out.println("File info: " + fileDetail.getName() + " " + fileDetail.getFileName() + " " + fileDetail.getSize());
System.out.println("Input size: " + uploadedInputStream.available());
```

**Comment by seshuad** [comment]
Here you go. this log is from trying to write to S3.

for a pdf file 

com.amazonaws.SdkClientException: Data read has a different length than the expected: dataLength=243499; expectedLength=252786; includeSkipped=false; in.getClass()=class com.amazonaws.internal.ReleasableInputStream; 

for a zip file

com.amazonaws.SdkClientException: Data read has a different length than the expected: dataLength=256116; expectedLength=270669; includeSkipped=false; in.getClass()=class com.amazonaws.internal.ReleasableInputStream;

**Comment by sapessi** [comment]
Thanks for the additional info. The fact that it works fine with text file suggests that perhaps the request body is not base64 encoded. Quick test for this by injecting the ContainerRequest in your method.

```java
public Response uploadFile(
        @QueryParam("tenantId") String tenantId,
        @QueryParam("poolId") String poolId,
        @QueryParam("vmname") String vmname,
        @QueryParam("logFileSize") long logFileSize,
        @FormDataParam("file") InputStream uploadedInputStream,
        @FormDataParam("file") FormDataContentDisposition fileDetail,
        @Context ContainerRequestContext req) {
    System.out.println("Is base64 encoded: " + ((AwsProxyHttpServletRequest)req.getProperty(JERSEY_SERVLET_REQUEST_PROPERTY)).getAwsProxyRequest().isBase64Encoded()
        );
...
}
```

**Comment by tedyu** [comment]
The output obtained from above statement is :

Is base64 encoded: false

**Comment by tedyu** [comment]
I modified getInputStream() method:
```
            log.info("body len: " + body.length());

            try {
                if (getContentType().startsWith("multipart/form-data")) {
                    bodyBytes = body.getBytes();
                } else {
                    if (encoding == null) {
                        log.info("Using ISO_8859 encoding");
                        encoding = StandardCharsets.ISO_8859_1.name();
                    }
                    bodyBytes = body.getBytes(encoding);
                }
            } catch (Exception e) {
                log.error("Could not read request with character encoding: " + SecurityUtils.crlf(encoding), e);
                bodyBytes = request.getBody().getBytes(StandardCharsets.ISO_8859_1.name());
            }

        log.info("Request size " + bodyBytes.length);
        ByteArrayInputStream requestBodyStream = new ByteArrayInputStream(bodyBytes);
        return new AwsServletInputStream(requestBodyStream, bodyBytes.length);
```
Here is the output:
```
[2018-12-28 07:43:35.525] 45ab77ea-0a74-11e9-9d0a-177d1431486c INFO c.v.v.x.c.p.i.s.AwsProxyHttpServletRequest - body len: 88366 
[2018-12-28 07:43:35.588] 45ab77ea-0a74-11e9-9d0a-177d1431486c INFO c.v.v.x.c.p.i.s.AwsProxyHttpServletRequest - Request size 159050 
```

**Comment by tedyu** [comment]
Here is AwsServletInputStream#read :
```
        @Override
        public int read()
                throws IOException {
            int readByte = bodyStream.read();
            bytesRead++;
            if (bodyStream.available() == 0) {
                log.info("has finally written " + bytesRead + " bytes with " + readByte);
                if (listener != null) {
                    listener.onAllDataRead();
                }
            }
            return readByte;
        }
```
Here is the output :
```
[2018-12-28 07:43:38.987] 45ab77ea-0a74-11e9-9d0a-177d1431486c INFO c.v.v.x.c.p.i.s.AwsProxyHttpServletRequest - has finally written 159050 bytes with 10 
[2018-12-28 07:43:38.988] 45ab77ea-0a74-11e9-9d0a-177d1431486c INFO c.v.v.x.c.p.i.s.AwsProxyHttpServletRequest - has finally written 159051 bytes with -1 
[2018-12-28 07:43:38.988] 45ab77ea-0a74-11e9-9d0a-177d1431486c INFO c.v.v.x.c.p.i.s.AwsProxyHttpServletRequest - has finally written 159052 bytes with -1 
```
I wonder why read() is called twice where return value is -1.
The outcome of the above change to getInputStream() method is that file uploaded to S3 is of size 0.

**Comment by tedyu** [comment]
Not sure it is related:
ReadListener is null for the AwsServletInputStream instance.

**Comment by tedyu** [comment]
Added the following two methods to AwsServletInputStream :
```
        @Override
        public void close() throws IOException {
            log.info("inside close");
            bodyStream.close();
        }

        @Override
        public int available() throws IOException {
            return length - bytesRead;
        }
```
However, there is no visible impact on the outcome of uploading the same pdf.
The log inside the close() method was not called.

**Comment by tedyu** [comment]
If I save the upload in a local file before calling putObject() to save on S3, the file size of the upload on S3 is no longer zero.
But the pdf file cannot be read since the length seen by AwsServletInputStream is much longer than the length of request body.

**Comment by tedyu** [comment]
I also tried StandardCharsets.UTF_16 and StandardCharsets.UTF_8 as parameter to getBytes().

StandardCharsets.UTF_16 resulted in 400 Bad Request - though the overridden close() was called:
```
[2018-12-28 08:32:34.210] 1b4cbe59-0a7b-11e9-b672-d51a6567bc66 INFO c.v.v.x.c.p.i.s.AwsProxyHttpServletRequest - has finally written 176734 bytes with 10 
[2018-12-28 08:32:34.229] 1b4cbe59-0a7b-11e9-b672-d51a6567bc66 INFO c.v.v.x.c.p.i.s.AwsProxyHttpServletRequest - has finally written 176735 bytes with -1 
[2018-12-28 08:32:34.229] 1b4cbe59-0a7b-11e9-b672-d51a6567bc66 INFO c.v.v.x.c.p.i.s.AwsProxyHttpServletRequest - has finally written 176736 bytes with -1 
[2018-12-28 08:32:34.286] 1b4cbe59-0a7b-11e9-b672-d51a6567bc66 INFO c.v.v.x.c.p.i.s.AwsProxyHttpServletRequest - inside close 
```
For UTF8, the behavior is the same as no-parameter getBytes().

**Comment by tedyu** [comment]
If I specify application/octet-stream as Content-Type in Postman, I got:

415 Unsupported Media Type

**Comment by tedyu** [comment]
From AwsProxyHttpServletRequest :
```
    private Map<String, Part> getMultipartFormParametersMap() {
        log.info("Entering getMultipartFormParameters");
```
The above method is called by :
```
    public Collection<Part> getParts() throws IOException, ServletException;
```
Looking at CloudWatch log, I don't see the 'Entering' line - though the content type is multipart/form-data.

**Comment by tedyu** [comment]
Thanks for the pointer.

I looked at the sample for using presigned URL.
Our setup is like this:
Some VMs may not have access to internet. They need to transfer the log bundle to a node with internet access.
It seems the presigned URL is useful for the node with internet access to upload log bundle to S3. However, we still need the multipart support for VMs to transfer log bundle to the node which has internet access.

Please let me know if my understanding is incorrect.

**Comment by tedyu** [comment]
When I tried to plugin the presigned URL approach into jersey resource I realized that the resource utilizes file upload feature (through MediaType.MULTIPART_FORM_DATA).
Without properly deciphering the multipart input, I cannot upload the file to S3 using presigned URL.

**Comment by sapessi** [comment]
I haven't been able to replicate the behavior in a unit test yet. Pretty sure I'm doing something wrong when creating the request object. I'll keep looking into it and hopefully push a fix for the next patch release.

**Comment by sapessi** [comment]
I've done some more digging into this @tedyu and @seshuad. I believe the root cause of the issue is described in [this StackOverflow answer](https://stackoverflow.com/questions/32426291/jersey-2-multipart-form-data-issue-inputstream-is-empty-available-0/33962167#33962167). TLDR is that Jersey doesn't have a built-in provider to read an `InputStream` from a multipart form. When I changed my test to use a `File` object instead it all worked fine.

```java
@Path("/file-size") @POST
@Consumes(MediaType.MULTIPART_FORM_DATA)
@Produces(MediaType.APPLICATION_JSON)
public Response fileSize(@FormDataParam("file") final File uploadedFile,
                         FormDataMultiPart multiPart,
                         @FormDataParam("file") FormDataContentDisposition fileDetail,
                         @Context ContainerRequestContext req) {
    SingleValueModel sv = new SingleValueModel();
    try {
        InputStream fileIs = new FileInputStream(uploadedFile);
        System.out.println("File: " + fileDetail.getName() + " " + fileDetail.getFileName() + " " + fileDetail.getSize());
        System.out.println("Size: " + fileIs.available());
        sv.setValue("" + fileIs.available());
        return Response.ok(sv).build();
    } catch (IOException e) {
        e.printStackTrace();
        return Response.status(500).build();
    }
}
```

**Comment by sapessi** [comment]
Closing since version `1.3.1` is out.
